src/main/java/io/quarkus/bot/util/PullRequestFilesMatcher.java

```diff
@@ -0,0 +1,74 @@
+package io.quarkus.bot.util;
+
+import com.hrakaroo.glob.GlobPattern;
+import com.hrakaroo.glob.MatchingEngine;
+import org.jboss.logging.Logger;
+import org.kohsuke.github.GHPullRequest;
+import org.kohsuke.github.GHPullRequestFileDetail;
+import org.kohsuke.github.PagedIterable;
+
+import java.util.Collection;
+
+public class PullRequestFilesMatcher {
+
+    private static final Logger LOG = Logger.getLogger(PullRequestFilesMatcher.class);
+
+    private final GHPullRequest pullRequest;
+
+    public PullRequestFilesMatcher(GHPullRequest pullRequest) {
+        this.pullRequest = pullRequest;
+    }
+
+    public boolean changedFilesMatchDirectory(Collection<String> directories) {
+        for (GHPullRequestFileDetail changedFile : pullRequest.listFiles()) {
+            for (String directory : directories) {
+
+                if (!directory.contains("*")) {
+                    if (changedFile.getFilename().startsWith(directory)) {
+                        return true;
+                    }
+                } else {
+                    try {
+                        MatchingEngine matchingEngine = GlobPattern.compile(directory);
+                        if (matchingEngine.matches(changedFile.getFilename())) {
+                            return true;
+                        }
+                    } catch (Exception e) {
+                        LOG.error("Error evaluating glob expression: " + directory, e);
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    public boolean changedFilesMatchFile(Collection<String> files) {
+
+        PagedIterable<GHPullRequestFileDetail> prFiles = pullRequest.listFiles();
+
+        if (prFiles == null || files == null) {
+            return false;
+        }
+
+        for (GHPullRequestFileDetail changedFile : prFiles) {
+            for (String file : files) {
+
+                if (!file.contains("*")) {
+                    if (changedFile.getFilename().endsWith(file)) {
```

Comment:
Hmmm, it used to be `startsWith` and you changed it to `endsWith`. I don't think that's what we want.

Reply:
I had to use `endsWith` because `getFilename` returns the fully qualified name of the file, including the path. So for directories it's `startsWith` and for files it's `endsWith`. Doing proper path parsing to extract the basename would be safer, though. (And on the directory path, doing path parsing to extract the path and drop the basename might be a good idea, except that it would change the existing semantics for `Triage`.) 

Looking at it again, I also see that the glob path is common across the directory and file methods. If it's truly common it should be extracted, but I suspect it shouldn't be common. At the very least, the basename should be passed in in the file path.

Reply:
I'm not sure I understand what you're saying. The existing code is working very well for pull requests triaging so I'm not sure why it became a problem.
Do you have an example of what's not working?

Reply:
The existing code still works fine, I just added a new requirement. 

Existing config: 

```
                                    triage:
                                      rules:
                                        - directories:
                                            - foo/*
                                            - bar/*
```
                                          
It matches on directories, using `startsWith()`.

Additional config:
```
                            workflows:
                                  rules:
                                    - allow:
                                        directories:
                                            - dir1
                                        files:
                                            - file1
```
It matches on directories, using `startsWith()`, and on files, using `endsWith()` (because to match  `someDir/someOtherDir/file1' against  'file1' you have to use `endsWith`). The `startsWith()` logic for directories is the same as what's used by the `triage` rule, so that codepath is shared across the two rules. It hasn't changed except by being pulled out to its own file. The `files` rule is unique to `workflows` so it's new code.

Reply:
That's where we disagree then. For matching a file that is a directory and not at the root, you have to use a glob pattern and make it `**/your-file`.

Reply:
So the above example would have to be: 

```
                            workflows:
                                  rules:
                                    - allow:
                                        directories:
                                            - ./dir1 [forgot the leading slash in the above example!]
                                            - **/file1
```

It seems not entirely user-friendly, and like it would need extra explanation in the docs. ("If you want to specify a filename, put it in the 'directories' rule with globbing.") At the moment, with what I've implemented, the above all works (exactly as it did before), but as a bonus, there's a more-concise `files:` option.